bam_stat.c

```diff
@@ -97,15 +97,129 @@ static void usage_exit(FILE *fp, int exit_status)
 {
     fprintf(fp, "Usage: samtools flagstat [options] <in.bam>\n");
     sam_global_opt_help(fp, "-.---@");
+    fprintf(fp, "  -O, --");
+    fprintf(fp, "output-fmt FORMAT[,OPT[=VAL]]...\n"
+            "               Specify output format (json, tsv)\n");
     exit(exit_status);
 }
 
+static void out_fmt_default(bam_flagstat_t *s)
+{
+    char b0[16], b1[16];
+    printf("%lld + %lld in total (QC-passed reads + QC-failed reads)\n", s->n_reads[0], s->n_reads[1]);
+    printf("%lld + %lld secondary\n", s->n_secondary[0], s->n_secondary[1]);
+    printf("%lld + %lld supplementary\n", s->n_supp[0], s->n_supp[1]);
+    printf("%lld + %lld duplicates\n", s->n_dup[0], s->n_dup[1]);
+    printf("%lld + %lld mapped (%s : %s)\n", s->n_mapped[0], s->n_mapped[1], percent(b0, s->n_mapped[0], s->n_reads[0]), percent(b1, s->n_mapped[1], s->n_reads[1]));
+    printf("%lld + %lld paired in sequencing\n", s->n_pair_all[0], s->n_pair_all[1]);
+    printf("%lld + %lld read1\n", s->n_read1[0], s->n_read1[1]);
+    printf("%lld + %lld read2\n", s->n_read2[0], s->n_read2[1]);
+    printf("%lld + %lld properly paired (%s : %s)\n", s->n_pair_good[0], s->n_pair_good[1], percent(b0, s->n_pair_good[0], s->n_pair_all[0]), percent(b1, s->n_pair_good[1], s->n_pair_all[1]));
+    printf("%lld + %lld with itself and mate mapped\n", s->n_pair_map[0], s->n_pair_map[1]);
+    printf("%lld + %lld singletons (%s : %s)\n", s->n_sgltn[0], s->n_sgltn[1], percent(b0, s->n_sgltn[0], s->n_pair_all[0]), percent(b1, s->n_sgltn[1], s->n_pair_all[1]));
+    printf("%lld + %lld with mate mapped to a different chr\n", s->n_diffchr[0], s->n_diffchr[1]);
+    printf("%lld + %lld with mate mapped to a different chr (mapQ>=5)\n", s->n_diffhigh[0], s->n_diffhigh[1]);
+}
+
+static void out_fmt_json(bam_flagstat_t *s) {
+    char b0[16], b1[16];
+    printf("{\n \"QC-passed reads\": { \n"
```

Comment:
I've a few more comments on the JSON output, now it's a bit easier to read (thanks).

As [JSON](https://www.json.org/) values can be numbers, I think it would be better to use them instead of strings.  It may help avoid unnecessary conversions, depending on what is reading the data.

For ease of parsing the values with percentages, it might be better to split them into separate entries, so instead of:
```json
{
  "mapped": "32 (94.12%)", 
}
```
it could be like this (note the `%` is dropped from the value to turn it into a number):
```json
{
   "mapped": 32,
   "mapped %": 94.12
}
```
This does leave the problem of what to do when `total` is zero so you can't calculate a percentage.  Currently it uses `N/A` which would turn the value into a string.  Possibly more convenient options might be to output `0.0` so it stays a number, or maybe `null` to indicate a missing value.  The `0.0` option could be easily done using:
```c
s->n_reads[0] ? (float)s->n_mapped[0] / s->n_reads[0] * 100.0 : 0.0
```
`null` or `"N/A"` would require a json-specific version of the `percent()` function.

Reply:
@daviesrob I see. So the idea is to have the JSON output to be as reliant to the spec as possible. That makes sense. With that, I suppose using `null` makes more sense even though for consistency among output formats 'N/A' might be more useful.